diff -Nru chillispot-1.0-coova.4/src/redir.c chillispot-1.0-coova.4.new/src/redir.c
--- chillispot-1.0-coova.4/src/redir.c	2007-01-31 13:37:55.000000000 +0000
+++ chillispot-1.0-coova.4.new/src/redir.c	2007-05-22 15:58:52.000000000 +0000
@@ -1453,9 +1453,7 @@
 	log_dbg("found %s, checking password", u);
 
 	if (conn->chap == 0) {
-	  int n;
-	  for (n=0; n < REDIR_MD5LEN; n++)
-	    tmp[n] = p[n] ^ chap_challenge[n];
+	  memcpy(tmp, p, REDIR_MD5LEN);
 	}
 	else if (conn->chap == 1) {
 	  MD5Init(&context);
@@ -1467,7 +1465,7 @@
 
 	tmp[REDIR_MD5LEN] = 0;
 
-	if (!memcmp(user_password, tmp, REDIR_MD5LEN)) 
+	if (!strncmp(user_password, tmp, REDIR_MD5LEN)) 
 	  match = 1; 
 
 	break;
